base/mpfr.jl

```diff
@@ -793,22 +833,27 @@ isfinite(x::BigFloat) = !isinf(x) && !isnan(x)
 @eval typemin(::Type{BigFloat}) = $(BigFloat(-Inf))
 
 function nextfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ptr{BigFloat}, Ptr{BigFloat}, Int32),
-          &z, &x, ROUNDING_MODE[end])
-    ccall((:mpfr_nextabove, :libmpfr), Int32, (Ptr{BigFloat},), &z) != 0
+    z = BigFloat(prec=precision(x))
+    ccall((:mpfr_set, :libmpfr), Int32, (Ptr{BigFloat}, Ptr{BigFloat}, Int32), &z, &x, ROUNDING_MODE[end])
+    ccall((:mpfr_nextabove, :libmpfr), Void, (Ptr{BigFloat},), &z) # modifies z
     return z
 end
 
 function prevfloat(x::BigFloat)
-    z = BigFloat()
-    ccall((:mpfr_set, :libmpfr), Int32, (Ptr{BigFloat}, Ptr{BigFloat}, Int32),
-          &z, &x, ROUNDING_MODE[end])
-    ccall((:mpfr_nextbelow, :libmpfr), Int32, (Ptr{BigFloat},), &z) != 0
+    z = BigFloat(prec=precision(x))
+    ccall((:mpfr_set, :libmpfr), Int32, (Ptr{BigFloat}, Ptr{BigFloat}, Int32), &z, &x, ROUNDING_MODE[end])
+    ccall((:mpfr_nextbelow, :libmpfr), Void, (Ptr{BigFloat},), &z) # modifies z
     return z
 end
 
-eps(::Type{BigFloat}) = nextfloat(BigFloat(1)) - BigFloat(1)
+function eps(x::BigFloat)
+    BigFloat(nextfloat(x) - x, prec=precision(x))
```

Comment:
You might be doing double rounding in this case, no? The result of the subtraction is calculated using the global precision, and then the result is converted down to precision of `x`.